azure/services/async/async.go

```diff
@@ -0,0 +1,137 @@
+	future := scope.GetLongRunningOperationState(resourceName, serviceName)
+	if future != nil {
+		return ProcessOngoingOperation(ctx, scope, client, resourceName, serviceName)
+	}
+
+	// No long running operation is active, so create the resource.
+	scope.V(2).Info("creating resource", "service", serviceName, "resource", resourceName, "resourceGroup", rgName)
+	sdkFuture, err := client.CreateOrUpdateAsync(ctx, spec)
+	if err != nil {
+		if sdkFuture != nil {
+			future, err := converters.SDKToFuture(sdkFuture, infrav1.PutFuture, serviceName, resourceName, rgName)
+			if err != nil {
+				return errors.Wrapf(err, "failed to create resource %s/%s (service: %s)", rgName, resourceName, serviceName)
+			}
+			scope.SetLongRunningOperationState(future)
+			return azure.WithTransientError(azure.NewOperationNotDoneError(future), retryAfter(sdkFuture))
+		}
+
+		return errors.Wrapf(err, "failed to create resource %s/%s (service: %s)", rgName, resourceName, serviceName)
+	}
+
+	scope.V(2).Info("successfully created resource", "service", serviceName, "resource", resourceName, "resourceGroup", rgName)
+	return nil
+}
+
+// DeleteResource implements the logic for deleting a resource Asynchronously.
+func DeleteResource(ctx context.Context, scope FutureScope, client Deleter, spec azure.ResourceSpecGetter, serviceName string) error {
+	resourceName := spec.ResourceName()
+	rgName := spec.ResourceGroupName()
+
+	// Check if there is an ongoing long running operation.
+	future := scope.GetLongRunningOperationState(resourceName, serviceName)
+	if future != nil {
+		return ProcessOngoingOperation(ctx, scope, client, resourceName, serviceName)
+	}
+
+	// No long running operation is active, so delete the resource.
+	scope.V(2).Info("deleting resource", "service", serviceName, "resource", resourceName, "resourceGroup", rgName)
+	sdkFuture, err := client.DeleteAsync(ctx, spec)
+	if err != nil {
+		if azure.ResourceNotFound(err) {
+			// already deleted
+			return nil
+		} else if sdkFuture != nil {
+			future, err := converters.SDKToFuture(sdkFuture, infrav1.DeleteFuture, serviceName, resourceName, rgName)
+			if err != nil {
+				return errors.Wrapf(err, "failed to delete resource %s/%s (service: %s)", rgName, resourceName, serviceName)
+			}
+			scope.SetLongRunningOperationState(future)
+			return azure.WithTransientError(azure.NewOperationNotDoneError(future), retryAfter(sdkFuture))
+		}
+
+		return errors.Wrapf(err, "failed to delete resource %s/%s (service: %s)", rgName, resourceName, serviceName)
+	}
+
+	scope.V(2).Info("successfully deleted resource", "service", serviceName, "resource", resourceName, "resourceGroup", rgName)
+	return nil
+}
+
+// retryAfter returns the max between the `RETRY-AFTER` header and the default requeue time.
+// This ensures we respect the retry-after header if it is set and avoid retrying too often during an API throttling event.
+func retryAfter(sdkFuture azureautorest.FutureAPI) time.Duration {
```

Comment:
@devigned PTAL

Reply:
`GetPollingDelay` is precisely what I was hoping would be used. You just added API informed throttling!!